IPIP/0000-ordered-car-responses.md

```diff
@@ -0,0 +1,236 @@
+however you MUST NOT stop the traversal because the same hash decoded two way may have different descent.
+
+For gateways:
+
+`Unordered` is the only MUST implement ordering.
+
+- `ProvenUsefullFirst`
+- `ForwardDFSWithSkippingOfDuplicates`
+- `ForwardDFSWithDuplicates`
+
+are SHOULD implement orderings.
+
+- `BackwardDFSWithSkippingOfDuplicates`
+- `BackwardDFSWithDuplicates`
+
+are MAY implement orderings.
+
+Adding `-OnlyUsefull` SHOULD be implemented by the gateway assuming they actually implement such feature.
+
+Implementing a stricter compatible ordering counts as ordering other compatible orderings.
+
+#### `Unordered`
+
+This is the current state of things and the default when `BlockOrder` requests.
+
+All blocks found doing a DFS search of all the roots must eventually be found in the file.
+
+Order is totally undefined and blocks might have duplicates as well as include blocks not part of the request.
+
+Altho you can have duplicates, this really meant to allows reverse proxies to cache more restrictive ones if they wish.
+Implementations SHOULD NOT emit duplicated blocks because this is a waste of bandwidth.
+
+##### `Unordered-OnlyUsefull`
+
+Because `Unordered` never cares about which blocks are received while reading, ensuring all blocks are useful would need to be finalized at the end by the client.
+
+#### `ProvenUsefullFirst`
+
+This is a slightly stricter version of `Unordered`.
+
+When a block is found in the stream but have not been proven useful yet, it MUST be ignored by clients (and thus the gateway MUST NOT assume the client have correctly received it).
+A block become proven useful when it has been seen as link of an other already proved useful block.
+The roots are the starting proven useful when starting the request.
+
+Note, the reason it is allowed but ignored to send useless blocks is for padding purposes, some implementations include small empty blocks in order to pad data to 4096.
+We want to be compatible with such implementations.
+
+We expect most concurrent oriented gateways to use this as their default because it keep bandwidth usage small and correspond to concurrent dag traversal patterns.
+
+The server SHOULD avoid sending the same blocks multiple times.
+
+#### `ForwardDFSWithSkippingOfDuplicates`, `BackwardDFSWithSkippingOfDuplicates`
+
+In this mode, DFS is precisely required either `Forward` or `Backward` (thoses indicates respectively weather maps and list should be traversed from low or high indexes first).
+
+In case a block that isn't precisely the next block in the DFS traversal is found it MUST be skipped by the client (and thus the gateway MUST NOT assume the client have correctly received it).
+
+The gateway SHOULD avoid sending the same blocks multiple times,
+the canonical way to do this is to create a seen CID set, add every CID traversed to it, then when traversing if the CID is in the set skip this branch completely.
```

Comment:
This is so a gateway is able to send a `ForwardDFSWithDuplicates` file to someone requesting a `ForwardDFSWithSkippingOfDuplicates` (because then when the gateway hits a duplicated part of the file the client is supposed to think this is padding data (or whatever reason someone would want to put useless blocks in car files) and ignore thoses unwanted blocks).

You add the `-OnlyUsefull` suffix, this is disallowed.